api/error_unexpected_response.go

```diff
@@ -0,0 +1,176 @@
+func NewUnexpectedResponseError(src UnexpectedResponseErrorSource, opts ...UnexpectedResponseErrorOption) UnexpectedResponseError {
+	new := src()
+	for _, opt := range opts {
+		opt(new)
+	}
+	if new.statusText == "" {
+		// the stdlib's http.StatusText function is a good place to start
+		new.statusFromCode(http.StatusText)
+	}
+
+	return *new
+}
+
+// Use textual representation of the given integer code. Called when status text
+// is not set using the WithStatusText option.
+func (e UnexpectedResponseError) statusFromCode(f func(int) string) {
+	e.statusText = f(e.statusCode)
+	if !e.HasStatusText() {
+		e.statusText = "unknown status code"
+	}
+}
+
+func (e UnexpectedResponseError) Error() string {
+	var eTxt strings.Builder
+	eTxt.WriteString("Unexpected response code")
+	if e.HasBody() || e.HasStatusCode() {
+		eTxt.WriteString(": ")
+	}
+	if e.HasStatusCode() {
+		eTxt.WriteString(fmt.Sprint(e.statusCode))
+		if e.HasBody() {
+			eTxt.WriteRune(' ')
+		}
+	}
+	if e.HasBody() {
+		eTxt.WriteString(fmt.Sprintf("(%s)", e.body))
+	}
+
+	if e.HasAdditional() {
+		eTxt.WriteString(fmt.Sprintf(". Additionally, an error occurred while constructing this error (%s); the body might be truncated or missing.", e.additional.Error()))
+	}
+
+	return eTxt.String()
+}
+
+// UnexpectedResponseErrorOptions are functions passed to NewUnexpectedResponseError
+// to customize the created error.
+type UnexpectedResponseErrorOption func(*UnexpectedResponseError)
+
+// WithError allows the addition of a Go error that may have been encountered
+// while processing the response. For example, if there is an error constructing
+// the gzip reader to process a gzip-encoded response body.
+func WithError(e error) UnexpectedResponseErrorOption {
+	return func(u *UnexpectedResponseError) { u.err = e }
+}
+
+// WithBody overwrites the Body value with the provided custom value
+func WithBody(b string) UnexpectedResponseErrorOption {
+	return func(u *UnexpectedResponseError) { u.body = b }
+}
+
+// WithStatusText overwrites the StatusText value the provided custom value
+func WithStatusText(st string) UnexpectedResponseErrorOption {
+	return func(u *UnexpectedResponseError) { u.statusText = st }
+}
+
+// WithExpectedStatuses provides a list of statuses that the receiving function
+// expected to receive. This can be used by API callers to provide more feedback
+// to end-users.
+func WithExpectedStatuses(s []int) UnexpectedResponseErrorOption {
+	return func(u *UnexpectedResponseError) {
+		u.expected = make([]int, len(s))
+		copy(u.expected, s)
+	}
+}
+
+// UnexpectedResponseErrorSource provides the basis for a NewUnexpectedResponseError.
+type UnexpectedResponseErrorSource func() *UnexpectedResponseError
+
+// FromHTTPResponse read an open HTTP response, drains and closes its body as
+// the data for the UnexpectedResponseError.
+func FromHTTPResponse(resp *http.Response) UnexpectedResponseErrorSource {
```

Comment:
The callsite of this would look like

```
err := api.FromHTTPResponse(resp)
```

which doesn't convey much about what the function does or returns - seems like this should have a name like `ErrorFromHTTPResponse` ?

Reply:
`FromHTTPResponse(resp)` doesn't return `error`, it makes a required function parameter that `NewUnexpectedResponseError(...)` consumes to populate the error it's building. The expectation is that you call `NewUnexpectedResponseError(FromHTTPResponse(resp))`. 

Is there a better conventional for naming the `UnexpectedResponseErrorSource` funcs? Also, since their is internal to the API package, maybe they should be unimported so Go API consumers don't think that they can/should use them.